### PR TITLE
Feature/biometrical

### PR DESCRIPTION
**Pull request by javiergonzper**
BUGS / iMPROVEMENTS 

- [x] (1) Passcode asked to switch touch id https://github.com/owncloud/ios-app/pull/54#issuecomment-396941343
- [x] (2) Redundant options when finger is not registered https://github.com/owncloud/ios-app/pull/54#issuecomment-396944841
- [X] (3) Biometrical removal from device point to a error message: https://github.com/owncloud/ios-app/pull/54#issuecomment-396986298 -> https://github.com/owncloud/ios-app/issues/80
- [x] (4) By disabling/enabling biometrical in device, it remains enabled in app https://github.com/owncloud/ios-app/pull/54#issuecomment-397268172

### CODE REVIEW

**Comment by javiergonzper** [comment]
@michaelstingl what do you think about the message that we show on the app when we will use the biometrical to unlock the device:

### For unlock the device using Touch ID

img_0077

"Unlock passcode".

### For unlock using Face ID

This message just appear the first time that the app try to use Face ID in order to explain to the user why we want to use the Face ID

img_0356

"Open the app without enter the passcode."

**Comment by michaelstingl** [comment]
I think it's okay 👍

File: ownCloud/Settings/Passcode/AppLockManager.swift [hunk 1]
```diff
@@ -263,4 +273,30 @@ class AppLockManager: NSObject {
...
+            if context.canEvaluatePolicy(LAPolicy.deviceOwnerAuthenticationWithBiometrics, error: nil) {
+                context.evaluatePolicy(LAPolicy.deviceOwnerAuthenticationWithBiometrics, localizedReason: "Unlock passcode".localized) { (success, _) in
+                    if success {
+                        //Fill the passcode dots
+                        DispatchQueue.main.async {
+                            self.passcodeViewController?.passcode = self.passcode
+                        }
+                        //Remove the passcode after small delay to give user feedback after use the biometrical unlock
+                        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
+                            self.dismissLockscreen(animated: true)
```
Comment (felix-schwarz) [inline ownCloud/Settings/Passcode/AppLockManager.swift]: Please use `attemptUnlock()` instead, which takes care of resetting/advancing internal state on retries, too.

File: ownCloud/Settings/Passcode/AppLockManager.swift [hunk 2]
```diff
@@ -263,4 +273,30 @@ class AppLockManager: NSObject {
...
+    func authenticateWithBiometrical() {
+
+        if  shouldDisplayLockscreen, biometricalSecurityEnabled {
+
+            let context = LAContext()
+
+            // Check if the device can evaluate the policy.
+            if context.canEvaluatePolicy(LAPolicy.deviceOwnerAuthenticationWithBiometrics, error: nil) {
+                context.evaluatePolicy(LAPolicy.deviceOwnerAuthenticationWithBiometrics, localizedReason: "Unlock passcode".localized) { (success, _) in
+                    if success {
```
Comment (felix-schwarz) [inline ownCloud/Settings/Passcode/AppLockManager.swift]: This codes lacks error handling. Especially for `LAError.code.biometryLockout`, which should be counted against failed attempts via `attemptUnlock(nil)`.

File: ownCloud/Settings/Passcode/AppLockManager.swift [hunk 3]
```diff
@@ -146,6 +151,11 @@ class AppLockManager: NSObject {
 			} else {
 				passcodeViewController?.screenBlurringEnabled = forceShow
 			}
+
+            // Show biometrical
+            if !forceShow {
+                authenticateWithBiometrical()
+            }
```
Comment (felix-schwarz) [inline ownCloud/Settings/Passcode/AppLockManager.swift]: In general, please make sure to use tabs in this project, so it matches the rest of the code. Right now it looks like this on GitHub and in Xcode when using a tab-width other than 4:

bildschirmfoto 2018-06-11 um 22 29 15

 (no worries, if you really prefer spaces, you can change tab/space setting on a per-Xcode-project and per-file level, too)

File: ownCloud/AppDelegate.swift [hunk 4]
```diff
@@ -39,7 +39,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 		window?.addSubview((navigationController?.view)!)
 		window?.makeKeyAndVisible()
 
-		AppLockManager.shared.showLockscreenIfNeeded()
+        AppLockManager.shared.showLockscreenIfNeeded()
```
[the remark this answers is not in the document]
Comment (javiergonzper) [reply ownCloud/AppDelegate.swift]: Looks like during a merge/rebase I had a conflict and I changed the tabs/spaces.
I changed it back.

File: ownCloud/Settings/Passcode/AppLockManager.swift (hunk 3 above)
Comment (javiergonzper) [reply ownCloud/Settings/Passcode/AppLockManager.swift]: No sorry, I prefer tabs. I do not know why that comment have spaces instead tabs. I will change it

File: ownCloud/Resources/Info.plist [hunk 5]
```diff
@@ -54,5 +54,7 @@
 		<string>UIInterfaceOrientationLandscapeLeft</string>
 		<string>UIInterfaceOrientationLandscapeRight</string>
 	</array>
+	<key>NSFaceIDUsageDescription</key>
+	<string>Open the app without enter the passcode.</string>
```
[the remark this answers is not in the document]
Comment (javiergonzper) [reply ownCloud/Resources/Info.plist]: 👍

File: ownCloud/Resources/en.lproj/InfoPlist.strings [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+ * Copyright (C) 2018, ownCloud GmbH.
+ *
+ * This code is covered by the GNU Public License Version 3.
+ *
+ * For distribution utilizing Apple mechanisms please see https://owncloud.org/contribute/iOS-license-exception/
+ * You should have received a copy of this license along with this program. If not, see <http://www.gnu.org/licenses/gpl-3.0.en.html>.
+ *
+ */
+
+"NSFaceIDUsageDescription" = "Open the app without enter the passcode.";
```
[the remark this answers is not in the document]
Comment (javiergonzper) [reply ownCloud/Resources/en.lproj/InfoPlist.strings]: 👍

File: ownCloud/Settings/Passcode/AppLockManager.swift [hunk 7]
```diff
@@ -98,6 +99,10 @@ class AppLockManager: NSObject {
 		}
 	}
 
+    var biometricalSecurityEnabled: Bool {
+        return self.userDefaults.bool(forKey: SecuritySettingsBiometricalKey)
+    }
```
[the remark this answers is not in the document]
Comment (javiergonzper) [reply ownCloud/Settings/Passcode/AppLockManager.swift]: 👍

File: ownCloud/Settings/Passcode/AppLockManager.swift [hunk 8]
```diff
@@ -263,4 +273,30 @@ class AppLockManager: NSObject {
 			}
 		}
 	}
+
+    // MARK: - Biometrical Unlock
+
+    func authenticateWithBiometrical() {
```
[the remark this answers is not in the document]
Comment (javiergonzper) [reply ownCloud/Settings/Passcode/AppLockManager.swift]: 👍

File: ownCloud/Settings/Passcode/AppLockManager.swift (hunk 1 above)
Comment (javiergonzper) [reply ownCloud/Settings/Passcode/AppLockManager.swift]: 👍

**Comment by jesmrec** [comment]
### (1)

Steps:

1. Enable Passcode in order to make visible touch id
2. Enable/Disable touch id

Current: Passcode is asked to enable/disable touch id

Why is asked the passcode to enable touch id? i think this is not needed for the touch id.

**Comment by jesmrec** [comment]
### (2)

Steps:

1. With touch id enabled, switch app to prompt touch id
2. Put a non-registered finger to display the error

Current: Two options displayed: `Enter Password` and `Cancel`. Both of them do the same action: dismiss the dialog and wait for the passcode.

Is this a predefined system dialog or could we avoid to show two redundant options?

**Comment by jesmrec** [comment]
### (3)

Steps:

1. Install the app
2. With a finger recorded, enable passcode and touch id
3. Go to device settings and remove all fingers

Current: every time the app is opened after switching or killing, the message `No entities are enrolled` is displayed, even though when the option for the touch id is not displayed in app settings anymore:

screen shot 2018-06-13 at 17 44 36

**Comment by felix-schwarz** [comment]
Regarding 1: asking for the passcode to enable or disable Face ID or Touch ID is common among apps. I guess its mostly about making sure
- the option is not enabled accidentally (giving other persons who have their fingerprint saved in Touch ID, but shouldn't have access to the app)
- the option is not enabled during the lock timeout by other persons who have their fingerprint saved in Touch ID, but shouldn't have access to the app.

Regarding 2: This is a pre-defined system dialog, but you're right in that the options are redundant. I've removed one of the options in the latest commits.

Regarding 3: Nice catch! Fixed that in the latest commits as well.

**Comment by jesmrec** [comment]
(1) -> not sure if it is often the fact that several people shares the same mobile device, beyond of testing purposes. It is not a bad behaviour, so let's wait for feedback of beta testers (sometimes they don't like to repeat same actions in a short time). On the other hand, enabling/disabling security set-up is not something to perform every hour.

(2) -> DONE

(3) -> i get still this issue (commit: https://github.com/owncloud/ios-app/pull/54/commits/14651371341ec4f10c17423983b3adb4b14d6094)

ezgif com-optimize 2

**Comment by jesmrec** [comment]
The discussion about (1) points me to this additional discussion (4):

Steps:

1. App has passcode + biometrical enabled
2. Remove from device settings the biometrical -> biometrical neither asked to enter the app nor displayed in app settings.
3. Go to device settings and register again fingers/face 
4. Open the app

Current: Biometrical is asked, because it was enabled in the app (step 1) in the moment it was removed from device settings (step 2)

Expected: two posibilities opened to discuss
- When app detects that biometrical stuff was removed from device, disabling the option (not only hiding it)
- Keep this behaviour, so the user wants the app protected by biometrical, regardless whether biometricals are removed and registered again.

**Comment by jesmrec** [comment]
@javiergonzper thanks for adding the first UI test. Only one tiny detail: the test is included together with the passcode tests, all of them into a folder called `PasscodeTests`. I would either change the name of such folder to `SecurityTests`, or create a new folder specific for the Biometrical tests, separate from passcode ones.

**Comment by felix-schwarz** [comment]
@jesmrec The current code change implemented this behaviour:
- if biometrical authentication is turned **on** and subsequently reset system-wide, the error message is shown
- if biometrical authentication is turned **off** and subsequently reset system-wide, no error message is shown (it was before)

I think the error message in the first case can be useful for support purposes. Imagine someone turning on biometrical authentication, then resetting Face ID / Touch ID (themselves or by others) and then contacting support the the option is turned on, but no biometric authentication is performed.

The error message then gives a clue as to what the issue is.

Automatically turning off biometrical authentication if it was reset is tempting, but I feel security settings should never be changed silently / automatically. So if we go down that path, an alert giving the user a "Keep enabled" / "Disable" choice (and requesting the passcode if "Disable" is chosen, just like in the settings) should be added.

**Comment by jesmrec** [comment]
> I think the error message in the first case can be useful for support purposes. Imagine someone turning on biometrical authentication, then resetting Face ID / Touch ID (themselves or by others) and then contacting support the the option is turned on, but no biometric authentication is performed.

Biometrical lock use to have a device support as background, so if the biometrical is removed from device, app should not show anything not supported (imho), so that the user can be confused about an error in the app that does point nowhere within the app itself. Make sense the scenario of supporting/testing only in case of a shared device, which i think is not usual in the oC app (maybe in others).

In any case, this is not a blocker and i will catch this:

> So if we go down that path, an alert giving the user a "Keep enabled" / "Disable" choice (and requesting the passcode if "Disable" is chosen, just like in the settings) should be added.

as a improvement to empower the security and users not to become confused.

Approved in my side.

**Pull request merged**